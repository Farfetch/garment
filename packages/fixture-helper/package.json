{
  "name": "@garment/fixture-helper",
  "license": "MIT",
  "version": "0.16.0",
  "main": "lib/index.js",
  "files": [
    "lib"
  ],
  "dependencies": {
    "@garment/logger": "^0.16.0",
    "@garment/runner": "^0.16.0",
    "@garment/workspace": "^0.16.0",
    "fs-extra": "8.1.0",
    "globby": "10.0.1",
    "strip-ansi": "^5.2.0",
    "tempy": "0.3.0"
  },
  "devDependencies": {
    "is-valid-path": "^0.1.1",
    "normalize-path": "^3.0.0"
  }
}
