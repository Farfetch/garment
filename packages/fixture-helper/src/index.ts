export * from './executeRunner';
export * from './initFixtureHelper';
export * from './resolver';
import normalizePath = require('normalize-path');
import isValidPath = require('is-valid-path');

const containsValidPath = (value: string): boolean =>
  value.split(/\s/).some((s) => isValidPath(s));

export const normalizeContainedPath = (value: string): string => {
  return value
    .split(/(\s)/)
    .map((s) => {
      if (isValidPath(s)) {
        return normalizePath(s);
      } else {
        return s;
      }
    })
    .join('');
};

export function replaceTestPath<T>(
  obj: T,
  testPath: string,
  replaceWith = '/test_path'
): T {
  const normalizedTestPathRegEx = new RegExp(normalizePath(testPath), 'g');
  return JSON.parse(
    JSON.stringify(obj, (_, value) =>
      typeof value === 'string' && containsValidPath(value)
        ? normalizeContainedPath(value).replace(
            normalizedTestPathRegEx,
            replaceWith
          )
        : value
    )
  );
}
