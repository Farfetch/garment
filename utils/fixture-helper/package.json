{
  "name": "@garment/fixture-helper",
  "license": "MIT",
  "version": "0.13.7",
  "main": "lib/index.js",
  "files": [
    "lib"
  ],
  "dependencies": {
    "@garment/logger": "^0.13.7",
    "@garment/runner": "^0.13.7",
    "@garment/workspace": "^0.13.7",
    "fs-extra": "8.1.0",
    "globby": "10.0.1",
    "strip-ansi": "^5.2.0",
    "tempy": "0.3.0"
  }
}
