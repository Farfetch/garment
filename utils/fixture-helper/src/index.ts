export * from './executeRunner';
export * from './initFixtureHelper';

export function replaceTestPath<T>(
  obj: T,
  testPath: string,
  replaceWith = '/test_path'
): T {
  return JSON.parse(
    JSON.stringify(obj, (_, value) =>
      typeof value === 'string'
        ? value.replace(new RegExp(testPath, 'g'), replaceWith)
        : value
    )
  );
}
