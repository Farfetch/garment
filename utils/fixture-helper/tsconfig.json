{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "rootDir": "src",
    "outDir": "lib",
    "composite": true
  },
  "exclude": [
    "node_modules",
    "lib",
    "__tests__"
  ],
  "references": [
    {
      "path": "../../core/logger"
    },
    {
      "path": "../../core/runner"
    },
    {
      "path": "../../core/workspace"
    }
  ]
}