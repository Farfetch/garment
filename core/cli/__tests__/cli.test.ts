import * as Path from 'path';
import execa = require('execa');
import stripAnsi from 'strip-ansi';
import { normalizeContainedPath } from '@garment/fixture-helper';

const cwd = Path.join(__dirname + '/fixtures/basic');
const garmentPath = Path.relative(cwd, 'core/cli/lib/cli.js');
const versionRegex = /v[0-9]+\.[0-9]+\.[0-9]+(.*)?/;
const noCacheArg = '--no-cache';

function execaSafe(...args: any[]) {
  //@ts-ignore
  return execa(...args)
    .then(({ stdout, stderr, ...rest }: any) => ({
      fulfilled: true,
      rejected: false,
      stdout: stripAnsi(stdout),
      stderr: stripAnsi(stderr),
      ...rest
    }))
    .catch((err: any) => ({
      fulfilled: false,
      rejected: true,
      reason: stripAnsi(err),
      stdout: '',
      stderr: stripAnsi(err.message)
    }));
}

test('cli should write help', async () => {
  const { stdout } = await execaSafe('node', [garmentPath, '-h', noCacheArg], {
    cwd: cwd
  });
  expect(stdout).toMatchSnapshot();
});

test('cli should execute if given correct number of commands', async () => {
  const { stdout } = await execaSafe(
    'node',
    [garmentPath, 'copy', noCacheArg],
    {
      cwd: cwd
    }
  );
  expect(stdout.replace(versionRegex, 'vX.X.X')).toMatchSnapshot();
});

test('cli should fail if given incorrect number of commands', async () => {
  const { stderr } = await execaSafe('node', [garmentPath, noCacheArg], {
    cwd: cwd
  });
  expect(normalizeContainedPath(stderr)).toMatchSnapshot();
});

test('cli should fail if given unexisting command', async () => {
  const { stderr } = await execaSafe('node', [garmentPath, 'dep', noCacheArg], {
    cwd: cwd
  });
  expect(normalizeContainedPath(stderr)).toMatchSnapshot();
});
