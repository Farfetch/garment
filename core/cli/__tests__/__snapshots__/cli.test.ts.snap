// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli should execute if given correct number of commands 1`] = `
"ðŸ§¤ Garment vX.X.X

 DONE  Project app 1 subtasks done

Projects: 1 total
Tasks:    1 total

"
`;

exports[`cli should fail if given incorrect number of commands 1`] = `
"Command failed with exit code 1: node ../../../lib/cli.js --no-cache
garment <cmd> [project]

Run command

Global Options:
  -h, --help     Show help                                             [boolean]
  -v, --version  Show version number                                   [boolean]
  --loglevel                                          [string] [default: \\"warn\\"]

Options:
  --watch          Activate watch mode                                 [boolean]
  --skipLifecycle  Prevent running pre- and post- lifecycle tasks      [boolean]
  --all            Run for all projects if none is provided            [boolean]
  --projects       Provide list of projects                              [array]
  --files          Provide list of files                                 [array]
  --exclude        Projects to exclude from run                          [array]
  --cache          Activate cache (--no-cache do deactivate)
                                                       [boolean] [default: true]
error Not enough non-option arguments: got 0, need at least 1"
`;

exports[`cli should fail if given unexisting command 1`] = `
"Command failed with exit code 1: node ../../../lib/cli.js dep --no-cache
error Specify one of available tasks: 
      
      copy
      go
      "
`;

exports[`cli should write help 1`] = `
"garment <cmd> [project]

Run command

Commands:
  garment <cmd> [project]                   Run command                [default]
  garment run <cmd> [project]               Run command
  garment generate [generatorName]          Run generator           [aliases: g]
  garment dep-graph                         Show projects dependency graph
  garment action-graph <task>               Show project task action graph
  [projects...]
  garment cache [sub-command]               Manage Garment cache
  garment init                              Initialize Garment workspace
  garment tasks [project]                   List project's tasks
  garment projects                          List projects

Global Options:
  -h, --help     Show help                                             [boolean]
  -v, --version  Show version number                                   [boolean]
  --loglevel                                          [string] [default: \\"warn\\"]

Options:
  --watch          Activate watch mode                                 [boolean]
  --skipLifecycle  Prevent running pre- and post- lifecycle tasks      [boolean]
  --all            Run for all projects if none is provided            [boolean]
  --projects       Provide list of projects                              [array]
  --files          Provide list of files                                 [array]
  --exclude        Projects to exclude from run                          [array]
  --cache          Activate cache (--no-cache do deactivate)
                                                       [boolean] [default: true]"
`;
